lib/rspec/matchers/built_in/base_matcher.rb

```diff
@@ -161,6 +161,32 @@ def improve_hash_formatting(inspect_string)
 
         include HashFormatting
 
+        # @private
+        module StringEncodingFormatting
+          # @api private
+          # @return [Boolean] True if the actual and expected string encoding are different.
+          #   i.e. the failure may be related to encoding differences and the encoding
+          #   should be shown to the user. false otherwise.
+          def string_encoding_differs?
+            actual.is_a?(String) && expected.is_a?(String) &&
+              # Older Ruby versions do not support #encoding
+              actual.respond_to?(:encoding) && expected.respond_to?(:encoding) &&
+              actual.encoding != expected.encoding
+          end
```

Comment:
We prefer to conditionally define things like this:
```suggestion
          if String.method_defined?(:encoding)
            def string_encoding_differs?
              actual.is_a?(String) && expected.is_a?(String) && actual.encoding != expected.encoding
            end
          else
            def string_encoding_differs?
              false
            end
          end
```